jax/_src/ops/scatter.py

```diff
@@ -66,7 +66,7 @@ def _scatter_update(x, idx, y, scatter_op, indices_are_sorted,
   """
 
   x = jnp.asarray(x)
-  y = jnp.asarray(y)
+  y = jnp.asarray(y, x.dtype)
```

Comment:
I don't think this is the correct fix. For example, if we write something like this:
```python
jnp.zeros(4, dtype='int32').at[0].set(1.5)
```
we do not want `1.5` silently cast to an integer type; we want a warning (that will eventually become an error)

Reply:
We could perhaps just read the original dtype of `y` to feed it into the warning/error logic...

Or do you have a better idea of the correct fix?

Reply:
The logic we need is "if `y` is a python scalar that can be safely cast to the dtype of `x`, then do so, but figure this out without attempting to cast `y` to the default dtype"

We don't have such a utility yet, but we could create one.

Reply:
Thanks!